### PR TITLE
[NOD-814] Migrate database to postgres + go-pg

### PR DESCRIPTION
**Pull request by dereeno**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,11 +3,11 @@ module github.com/kaspanet/faucet
 go 1.13
 
 require (
+	github.com/go-pg/pg/v9 v9.1.3
```
Comment (stasatdaglabs) [inline go.mod:6]: Please run go mod tidy to make sure gorm is truly gone.

File: ip_usage.go [hunk 2]
```diff
@@ -32,16 +32,11 @@ func validateIPUsage(r *http.Request) error {
...
 	ip, err := ipFromRequest(r)
 	if err != nil {
 		return err
 	}
-	dbResult := db.Model(&ipUse{}).Where(&ipUse{IP: ip}).Where("last_use BETWEEN ? AND ?", timeBeforeMinRequestInterval, now).Count(&count)
-	dbErrors := dbResult.GetErrors()
-	if httpserverutils.HasDBError(dbErrors) {
-		return httpserverutils.NewErrorFromDBErrors("Some errors were encountered when checking the last use of an IP:", dbResult.GetErrors())
-	}
+	count, err := db.Model(&ipUse{}).Where("ip = ?", ip).Where("last_use BETWEEN ? AND ?", timeBeforeMinRequestInterval, now).Count()
```
Comment (stasatdaglabs) [inline ip_usage.go]: Please split this line to multiple lines.

File: database/database.go [hunk 3]
```diff
@@ -3,43 +3,31 @@ package database
...
+	"github.com/go-pg/pg/v9"
 	"os"
 
 	"github.com/pkg/errors"
 
 	"github.com/golang-migrate/migrate/v4/source"
-	"github.com/jinzhu/gorm"
 	"github.com/kaspanet/faucet/config"
 
 	"github.com/golang-migrate/migrate/v4"
```
Comment (stasatdaglabs) [inline database/database.go]: Please gofmt the project to tidy up the imports.

**Review by stasatdaglabs — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (dereeno) [reply go.mod:6]: Done already

**Review by stasatdaglabs — APPROVED** [review]

File: config/config.go [hunk 4]
```diff
@@ -36,7 +38,9 @@ type Config struct {
 	HTTPListen  string  `long:"listen" description:"HTTP address to listen on (default: 0.0.0.0:8081)"`
 	KasparovURL string  `long:"kasparov-url" description:"The kasparov url to connect to"`
 	PrivateKey  string  `long:"private-key" description:"Faucet Private key"`
-	DBAddress   string  `long:"dbaddress" description:"Database address"`
+	DBHost      string  `long:"dbhost" description:"Database host"`
```
Comment (svarogg) [inline config/config.go]: Please add defaults in description

File: config/config.go [hunk 5]
```diff
@@ -36,7 +38,9 @@ type Config struct {
...
+	DBPort      string  `long:"dbport" description:"Database port"`
+	DBSSLMode   string  `long:"dbsslmode" description:"Database SSL mode {disable, allow, prefer, require, verify-ca, verify-full}"`
```
Comment (svarogg) [inline config/config.go]: You can specify the choices like so:
```
	// Example of a flag restricted to a pre-defined set of strings
	Animal string `long:"animal" choice:"cat" choice:"dog"`
```

File: ip_usage.go [hunk 6]
```diff
@@ -32,16 +32,15 @@ func validateIPUsage(r *http.Request) error {
...
 	ip, err := ipFromRequest(r)
 	if err != nil {
 		return err
 	}
-	dbResult := db.Model(&ipUse{}).Where(&ipUse{IP: ip}).Where("last_use BETWEEN ? AND ?", timeBeforeMinRequestInterval, now).Count(&count)
-	dbErrors := dbResult.GetErrors()
-	if httpserverutils.HasDBError(dbErrors) {
-		return httpserverutils.NewErrorFromDBErrors("Some errors were encountered when checking the last use of an IP:", dbResult.GetErrors())
-	}
+	count, err := db.Model(&ipUse{}).
```
Comment (svarogg) [inline ip_usage.go:39]: I think you are ignoring err here.

File: ip_usage.go [hunk 7]
```diff
@@ -58,10 +57,15 @@ func updateIPUsage(r *http.Request) error {
...
 	}
-	dbResult := db.Where(&ipUse{IP: ip}).Assign(&ipUse{LastUse: time.Now()}).FirstOrCreate(&ipUse{})
-	dbErrors := dbResult.GetErrors()
-	if httpserverutils.HasDBError(dbErrors) {
-		return httpserverutils.NewErrorFromDBErrors("Some errors were encountered when upserting the IP to the new date:", dbResult.GetErrors())
+
+	ipUse := &ipUse{IP: ip}
+	_, err = db.Model(ipUse).
+		OnConflict("(ip) DO UPDATE").
+		Set("last_use = ?", time.Now()).
```
Comment (svarogg) [inline ip_usage.go]: Just making sure: the Set will be triggered both on conlfict and not on conflict, right?
Maybe we want to set last_use inside `ipUse` to start with, to be more explicit either way?

**Review by svarogg — CHANGES_REQUESTED** [review]

File: ip_usage.go (hunk 7 above)
Comment (dereeno) [reply ip_usage.go]: Yes, it will set it either way. But your suggestion is more elegant, fixed.

**Review by svarogg — APPROVED** [review]

**Pull request merged**